### PR TITLE
Allow shell support to work when set -u is set

### LINKED ISSUE #3709: Cannot find any package installation
I decided to upgrade my spack and nothing works anymore.
So, I decided to have a fresh installation and got the latest tag 0.10.0.

```
$ spack find -p 
==> 4 installed packages.
-- linux-rhel6-x86_64 / gcc@4.6.1 -------------------------------
    libffi@3.2.1       /usr/workspace/wsa/bluenet/thiago/spack/opt/spack/linux-rhel6-x86_64/gcc-4.6.1/libffi-3.2.1-dtonmgxweviawxhaxh5ksuoyzjoprq7u
    ruby@2.2.0         /usr/workspace/wsa/bluenet/thiago/spack/opt/spack/linux-rhel6-x86_64/gcc-4.6.1/ruby-2.2.0-qezm2l32dhzutgy7eqnfec3tk2rjkliv
    tmuxinator@0.6.11  /usr/workspace/wsa/bluenet/thiago/spack/opt/spack/linux-rhel6-x86_64/gcc-4.6.1/tmuxinator-0.6.11-75czzxglwibzesozshh4tfev3v5ryrfs
    zlib@1.2.10        /usr/workspace/wsa/bluenet/thiago/spack/opt/spack/linux-rhel6-x86_64/gcc-4.6.1/zlib-1.2.10-dwavb3pn4os4kosd2zneb5bpjzpkeygn
```
However, it cannot find any package install:

```
$ spack use tmuxinator
Prepending: tmuxinator-0.6.11-gcc-4.6.1-75czzxg (not found)
```
Thanks

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #3709.

This is a continuation of #2417 and #2418. I don't actually use this script, so can someone who does test these changes? I'm still not convinced that everything works if you install Spack in a directory whose path contains a space.

### CODE REVIEW

**Comment by tgamblin** [comment]
@adamjstewart: looks like this needs a rebase.  I am not sure how best to test this, but could we ensure that it works by *always* executing the shell support in a `-u` shell?  e.g., we could wrap `setup-env.sh` in [something like this](https://stackoverflow.com/questions/15656492/linux-shell-try-catch-finally) and always `set -u` at the beginning, then restore it to whatever it was before at the end.

@thiagotei: can you test this out once @adamjstewart has rebased?

**Comment by adamjstewart** [comment]
Don't think it needs a rebase anymore now that the conflicting PR was reverted.

We can always set -u, although it's kind of annoying to maintain compatibility. Does set -u work in other shells? I've only ever used Bash so I haven't been thinking about POSIX compatibility.

**Comment by adamjstewart** [comment]
P.S. I don't even use `setup-env.sh` so it doesn't matter to me.

**Comment by thiagotei** [comment]
@tgamblin  I can test it! Just let me know where to pull from.

**Comment by adamjstewart** [comment]
@thiagotei Try:
```
$ git remote add adamjstewart https://github.com/adamjstewart/spack.git
$ git fetch adamjstewart fixes/set-u
$ git checkout fixes/set-u
```

**Comment by thiagotei** [comment]
@adamjstewart with `set -u` I did not get error, but I got a different output than when using `set +u`.
With `set -u`:
```bash
thiago@cab669 ~$ spack
usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-s] [-V] SUBCOMMAND ...
spack: error: argument SUBCOMMAND: invalid choice: '' choose from:
    activate  bootstrap  cd        clean     compilers  configure  deactivate  dependents  doc   env         fetch  flake8  help  install  load      md5     module  pkg        purge   reindex  restage  spec   test       unload  url  versions
    arch      build      checksum  compiler  config     create     debug       diy         edit  extensions  find   graph   info  list     location  mirror  patch   providers  python  repo     setup    stage  uninstall  unuse   use  view
```

With `set +u`:
```bash
thiago@cab669 ~$ spack
usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-s] [-V] SUBCOMMAND ...

Spack: the Supercomputing PACKage Manager.

spec expressions:
  PACKAGE [CONSTRAINTS]

    CONSTRAINTS:
      @version
      %compiler  @compiler_version
      +variant
      -variant or ~variant
      =architecture
      [^DEPENDENCY [CONSTRAINTS] ...]

positional arguments:
  SUBCOMMAND
    activate        activate a package extension
    arch            print architecture information about this machine
    bootstrap       create a new installation of spack in another prefix
    build           stops at build stage when installing a package, if possible
    cd              cd to spack directories in the shell
    checksum        checksum available versions of a package
    clean           remove build stage and source tarball for packages
    compiler        manage compilers
    compilers       list available compilers, same as 'spack compiler list'
    config          get and set configuration options
    configure       stops at configuration stage when installing a package, if possible
    create          create a new package file
    deactivate      deactivate a package extension
    debug           debugging commands for troubleshooting Spack
    dependents      show installed packages that depend on another
    diy             do-it-yourself: build from an existing source directory
    doc             run pydoc from within spack
    edit            open package files in $EDITOR
    env             run a command with the install environment for a spec
    extensions      list extensions for package
    fetch           fetch archives for packages
    find            find installed spack packages
    flake8          runs source code style checks on Spack. requires flake8
    graph           generate graphs of package dependency relationships
    help            get help on spack and its commands
    info            get detailed information on a particular package
    install         build and install packages
    list            print available spack packages to stdout in different formats
    load            add package to environment using modules
    location        print out locations of various directories used by Spack
    md5             calculate md5 checksums for files/urls
    mirror          manage mirrors
    module          manipulate module files
    patch           patch expanded archive sources in preparation for install
    pkg             query packages associated with particular git revisions
    providers       list packages that provide a particular virtual package
    purge           remove temporary build files and/or downloaded archives
    python          launch an interpreter as spack would launch a command
    reindex         rebuild Spack's package database
    repo            manage package source repositories
    restage         revert checked out package source code
    setup           create a configuration script and module, but don't build
    spec            print out abstract and concrete versions of a spec
    stage           expand downloaded archive in preparation for install
    test            a thin wrapper around the pytest command
    uninstall       remove an installed package
    unload          remove package from environment using module
    unuse           remove package from environment using dotkit
    url             debugging tool for url parsing
    use             add package to environment using dotkit
    versions        list available versions of a package
    view            produce a single-rooted directory view of a spec

optional arguments:
  -h, --help        show this help message and exit
  -d, --debug       write out debug logs during compile
  -D, --pdb         run spack under the pdb debugger
  -k, --insecure    do not check ssl certificates when downloading
  -m, --mock        use mock packages instead of real ones
  -p, --profile     profile execution using cProfile
  -v, --verbose     print additional output during builds
  -s, --stacktrace  add stacktrace information to all printed statements
  -V, --version     show program's version number and exit
```

**Comment by adamjstewart** [comment]
@citibeth Didn't you report something like this as well? I'll take a look.

**Comment by adamjstewart** [comment]
This doesn't actually have anything to do with `set -u`, it's a bug introduced in my changes. I'll fix it.

**Comment by thiagotei** [comment]
@adamjstewart  `set -u\+u` was the only thing that I changed.

**Comment by adamjstewart** [comment]
Pretty sure the problem is my use of quotes around some of the arguments. This can be reproduced without `setup-env.sh` by doing the following:
```
$ spack
# works
$ spack ''
# fails
```

**Comment by adamjstewart** [comment]
Fixed. I can take a look at always using `set -u`, possibly in all of our other shell scripts as well, but it would take some time.

**Comment by tgamblin** [comment]
@adamjstewart: it's not the highest priority thing.  It might help the shell code but there are bigger problems :)

**Comment by adamjstewart** [comment]
Yeah, about 450 of them. I've been trying to slog through them but a lot are unresolved discussions or package build problems. The biggest remaining issues are concretization, which you're working on, and external package configuration, which is a mess.

File: share/spack/setup-env.sh [hunk 1]
```diff
@@ -58,38 +58,38 @@
 
 function spack {
     # Zsh does not do word splitting by default, this enables it for this function only
-    if [ -n "${ZSH_VERSION:-}" ]; then
+    if [[ "${ZSH_VERSION:-}" ]]; then
```
Comment (junghans) [inline share/spack/setup-env.sh]: @adamjstewart : In `[[ ... ]]` quoting of variables isn't necessary. (In `[ ... ]` it is!)

Comment (adamjstewart) [inline share/spack/setup-env.sh]: I've always used both obsessively lol.

**Comment by adamjstewart** [comment]
Closing as this introduces a lot of Bashisms and will conflict with the attempt to make the script more POSIX compatible in #4048.

**Pull request closed without merging**